### PR TITLE
Added lifecycle event caching + conflict dialog

### PR DESCRIPTION
**Pull request by Mrez95**
**Changes**
- Fixed caching issues by introducing proper event lifecycles
- Conflict detection 
- Conflict dialog 

**Easiest Test Case**
- Add `ECE 106` + `ECE 224`

**When adding a conflicting course**
image

### CODE REVIEW

File: app/src/main/java/com/example/peterchu/watplanner/coursedetail/CourseDetailPresenter.java [hunk 1]
```diff
@@ -85,9 +92,21 @@ public void onFailure() {
                     });
             isFirstLoad = false;
         }
+        // retrieve from cache.
+        schedule = dataRepository.getCourseSchedules();
+    }
+
+    public void pause() {
+        // save to cache.
+        dataRepository.setCourseSchedules(schedule);
```
Comment (yongla) [inline app/src/main/java/com/example/peterchu/watplanner/coursedetail/CourseDetailPresenter.java:101]: Why do we have to call:
```
schedule = dataRepository.getCourseSchedules();
... // and then
dataRepository.setCourseSchedules(schedule);
```

in the course details page?

Comment (Mrez95) [reply app/src/main/java/com/example/peterchu/watplanner/coursedetail/CourseDetailPresenter.java:101]: cause when users add/remove a course in this page, the resulting schedule needs to be cached. Also, conflict detection happens in this page.

Comment (yongla) [reply app/src/main/java/com/example/peterchu/watplanner/coursedetail/CourseDetailPresenter.java:101]: aight, makes sense 💯

**Review by yongla — APPROVED** [review]

**Review by PeterChu2 — APPROVED** [review]
let's ship!! so excited

**Pull request merged**